atlasdb-dbkvs/src/main/java/com/palantir/atlasdb/keyvalue/dbkvs/impl/DbKvs.java

```diff
@@ -624,46 +629,57 @@ private ColumnSelection getColumnSelectionForRange(RangeRequest range) {
             }
         }
 
-        ColumnSelection columns = ColumnSelection.all();
-        if (!range.getColumnNames().isEmpty()) {
-            columns = ColumnSelection.create(range.getColumnNames());
-        }
+        ColumnSelection columns = range.getColumnNames().isEmpty()
+                ? ColumnSelection.all() : ColumnSelection.create(range.getColumnNames());
 
-        SetMultimap<Cell, Long> results = getTimestampsByCell(table, rows, columns, timestamp);
+        ResultWithToken results = getTimestampsByCell(table, rows, columns, timestamp, batchSize, range.isReverse());
 
         NavigableMap<byte[], SortedMap<byte[], Set<Long>>> cellsByRow =
-                Cells.breakCellsUpByRow(Multimaps.asMap(results));
-        if (range.isReverse()) {
-            cellsByRow = cellsByRow.descendingMap();
-        }
+                Cells.breakCellsUpByRow(Multimaps.asMap(results.results));
+
         List<RowResult<Set<Long>>> finalResults = Lists.newArrayListWithCapacity(results.size());
         for (Entry<byte[], SortedMap<byte[], Set<Long>>> entry : cellsByRow.entrySet()) {
             finalResults.add(RowResult.create(entry.getKey(), entry.getValue()));
         }
-        byte[] nextRow = null;
-        boolean mayHaveMoreResults = false;
-        byte[] lastRow = range.isReverse() ? rows.first() : rows.last();
-        if (!RangeRequests.isTerminalRow(range.isReverse(), lastRow)) {
-            nextRow = RangeRequests.getNextStartRow(range.isReverse(), lastRow);
-            mayHaveMoreResults = rows.size() == maxRows;
+        byte[] nextRow = results.currentRow;
+        boolean mayHaveMoreResults = !results.finishedRange;
```

Comment:
I would like to see this logic broken out into a place where it's unit testable, as it's quite complex.